### ISSUE TITLE
MDBX put is dramatically slower in debug builds compared to LMDB

### ISSUE THREAD

**Issue by Reisse** [leo-yuriev/libmdbx#80]
I have the following simple piece of code, which fills the database with some test data:

```c
#include "mdbx.h"

int main(int argc,char * argv[])
{
    int rc;
    MDBX_env *env;
    MDBX_dbi dbi;
    MDBX_val key, data;
    MDBX_txn *txn;
    int kval;
    char sval[1024] = {0}; // fine to fill with zeroes

    rc = mdbx_env_create(&env);
    rc = mdbx_env_set_mapsize(env, 10ULL * 1024 * 1024 * 1024); // 10 GiB
    rc = mdbx_env_open(env, "./testdb", 0, 0664);
    rc = mdbx_txn_begin(env, NULL, 0, &txn);
    rc = mdbx_dbi_open(txn, NULL, 0, &dbi);

    key.iov_len = sizeof(int);
    key.iov_base = &kval;
    data.iov_len = sizeof(sval);
    data.iov_base = sval;

    for (kval = 0; kval < 500'000; ++kval) {
        rc = mdbx_put(txn, dbi, &key, &data, 0);
    }
    rc = mdbx_txn_commit(txn);

    mdbx_dbi_close(env, dbi);
    mdbx_env_close(env);
    return 0;
}
```

It takes ~10 minutes to execute on my machine when compiled with the flags `-m64 -std=gnu++17 -fno-omit-frame-pointer -ffast-math -fno-finite-math-only -pthread -march=core2 -mtune=corei7 -g` (hereinafter "debug"). `perf` shows that MDBX spent majority of time in `mdbx_dpl_search`. The same code takes ~four seconds to execute when `-O3 -flto=full` (hereinafter "release") flags are added.

When I rewrite this code to use LMDB, it takes ~10 seconds to execute in debug build, and ~5 seconds to execute in release build. Could you please clarify why the difference is so huge here?

**Comment by leo-yuriev** [comment]
This is neither a bug nor a problem, but intentionally behavior.
MDBX used a lot of assertions which are enabled while library builded without `-DNDEBUG`.
So, you must decide whether you need a debug build with additional checks or a release build with high performance.

**Comment by leo-yuriev** [comment]
@Reisse, could we close this issue?

**Comment by Reisse** [comment]
Could you give me one day please? I've tried running with NDEBUG once and have a feeling that setting it doesn't help, but didn't yet had time to test it thoroughly.

**Comment by leo-yuriev** [comment]
@Reisse, FYI, with modern GCC/CLANG the [` -Og`](https://www.phoronix.com/scan.php?page=news_item&px=MTE3NDg) option should be useful in your case.

**Comment by Reisse** [comment]
Yes, thanks, `-DNDEBUG -Og` helps, but `-DNDEBUG` without optimization doesn't.

The reason of such slowness is this check in `SEARCH_IMPL` macro:
```c
 for (TYPE_LIST *scan = begin; scan < first; ++scan)                        \
   assert(CMP(*scan, item));                                                \
 for (TYPE_LIST *scan = first; scan < end; ++scan)                          \
   assert(!CMP(*scan, item));                                               \
```
Amirite that `SEARCH_IMPL` does binary search first, then linearly checks that the range is sorted, and that correct element is found?

Even when `NDEBUG` effectively eliminates `assert(...)`, non-optimized empty for-loop penalty is enough to make MDBX orders of magnitude slower than LMDB. So the only way to get reasonable performance is to enable at least some optimizations, for the compiler to be able to dump empty for-loop.

---

I should say that now I strongly disagree with closing an issue as invalid. I'm all in for assertions and sanity checks, but changing `O(log n)` algo to `O(n)` in one of the hottest places in the library is ridiculous. It renders MDBX completely unusable in debug mode.

Removing this particular check brings MDBX debug build performance back to LMDB level even without `NDEBUG`. Are the benefits of this check enough to keep it on by default? I suggest making it optional in some "fortified" mode with extended assertions or adding a sane limit on how much of the range is scanned. That is, instead of checking everything from `begin` to `end`, iterate from `max(first - X, begin)` to `min(first + X, end)`, where X is reasonably small.

**Comment by leo-yuriev** [comment]
1. I think there is a misunderstanding about the purpose of "debug builds". Primary it is to debug library itself, but not an applications. But a debug build _can also be_ useful when debugging an application, since there is often no way to determine where the bug is in an application code or in the library.

2. There is often confusion between a "Build with debug information" and a "Debug build", also called a "Checked build" (for instance read about  [Windows Checked build](https://docs.microsoft.com/en-us/windows-hardware/drivers/devtest/checked-build-of-windows)). Anyway the "Debug build" of libmdbx means the "Checked build" with a lot of additional checks.

3. A debug build _MUST do a lot of checks_, including the most stupid ones, because that's its purpose. Therefore, you should not complain about its performance especially if you do not use compiler optimization.

Nonetheless, for practical reasons, libmdbx support several levels of checking. So I will consider to "move" some checks to a more strict level.

**Comment by Reisse** [comment]
Sorry for the delayed response.

>     1. I think there is a misunderstanding about the purpose of "debug builds". Primary it is to debug library itself, but not an applications.

That's exactly what I was doing.

I found [the bug in LMDB](https://www.openldap.org/its/index.cgi/Software%20Bugs?id=9155) and wanted to check if it also could affect MDBX. Skimming quickly through the commit history, I didn't found related fixes in MDBX source code, so I decided to check it with the same test I used for LMDB. Of course I wanted to have default debug assertions enabled in case something goes wrong, but also I expected my test to finish in reasonable time. Otherwise, how to fight bugs affecting only large databases/transactions if it's not possible to work with them in debug builds?

In the end, the test passed without ASAN warnings on MDBX, so looks like everything is fine here :)

>     3. A debug build MUST do a lot of checks, including the most stupid ones, because that's its purpose. Therefore, you should not complain about its performance especially if you do not use compiler optimization.

I'd like to state again that the problem described the issue persisted with `-NDEBUG` flag enabled, thus, it was specific not only to debug mode, but also to non-optimized builds. So moving from "debug (checked) build" to "non-optimized build with debug information enabled" didn't help.

> Nonetheless, for practical reasons, libmdbx support several levels of checking. So I will consider to "move" some checks to a more strict level.

Thanks!
